File: lib/python/secrets_sdk/rotation.py
"""Secret rotation helpers.

Provides rotation policy definitions, age checking for Vault secrets,
and SOPS key rotation utilities.
"""

from __future__ import annotations

import logging
import re
import subprocess
import shutil
from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import TYPE_CHECKING, Any

from secrets_sdk.exceptions import RotationError, SopsNotInstalledError
from secrets_sdk.models import AgeReport

if TYPE_CHECKING:
    from secrets_sdk.vault import VaultClient

logger = logging.getLogger(__name__)


@dataclass
class RotationPolicy:
    """Defines rotation requirements for a class of secrets.

    Attributes:
        name: Human-readable policy name (e.g., "database-creds", "api-keys").
        max_age_days: Maximum allowed age in days before rotation is required.
        paths: Vault KV paths this policy applies to (glob-like patterns).
        warn_age_days: Age in days at which to start warning (default: 80% of max).
        auto_rotate: Whether the system should attempt automatic rotation.
        notify_channels: Where to send rotation notifications (e.g., ["slack", "email"]).
    """

    name: str
    max_age_days: float = 90.0
    paths: list[str] = field(default_factory=list)
    warn_age_days: float = 0.0
    auto_rotate: bool = False
    notify_channels: list[str] = field(default_factory=list)

    def __post_init__(self) -> None:
        if self.warn_age_days <= 0:
            self.warn_age_days = self.max_age_days * 0.8

    def matches_path(self, path: str) -> bool:
        """Check if a Vault path matches any of this policy's path patterns.

        Supports simple glob: `*` matches any single path segment,
        `**` is not supported (use multiple patterns instead).

        Args:
            path: Vault secret path to check.

        Returns:
            True if the path matches any pattern.
        """
        for pattern in self.paths:
            regex = _glob_to_regex(pattern)
            if re.match(regex, path):
                return True
        return False


# Default rotation policies for common secret types
DEFAULT_POLICIES: list[RotationPolicy] = [
    RotationPolicy(
        name="database-credentials",
        max_age_days=90,
        paths=["kv/data/*/database/*", "kv/data/*/db/*"],
    ),
    RotationPolicy(
        name="api-keys",
        max_age_days=180,
        paths=["kv/data/*/api-keys/*", "kv/data/*/apikeys/*"],
    ),
    RotationPolicy(
        name="service-accounts",
        max_age_days=365,
        paths=["kv/data/*/service-accounts/*"],
    ),
    RotationPolicy(
        name="tls-certificates",
        max_age_days=90,
        paths=["kv/data/*/certs/*", "kv/data/*/tls/*"],
    ),
    RotationPolicy(
        name="ssh-keys",
        max_age_days=90,
        paths=["kv/data/*/ssh/*"],
    ),
]


def check_secret_age(
    vault_client: "VaultClient",
    path: str,
    max_age_days: float = 90.0,
    kv_mount: str | None = None,
) -> AgeReport:
    """Check the age of a KV v2 secret and determine if rotation is needed.

    Args:
        vault_client: Authenticated VaultClient instance.
        path: Secret path (without mount prefix).
        max_age_days: Maximum allowed age in days.
        kv_mount: KV mount point override. Uses client default if None.

    Returns:
        AgeReport with age details and rotation recommendation.

    Raises:
        RotationError: If the metadata cannot be retrieved.
    """
    from secrets_sdk.exceptions import VaultSecretNotFound, VaultConnectionError

    try:
        metadata = vault_client.kv_metadata(path)
    except VaultSecretNotFound:
        return AgeReport(
            path=path,
            max_age_days=max_age_days,
            needs_rotation=False,
            detail=f"Secret not found at path: {path}",
        )
    except VaultConnectionError as exc:
        raise RotationError(f"Cannot check age of {path}: {exc}")

    now = datetime.now(timezone.utc)
    age_days = 0.0
    created = metadata.created_time

    if created is not None:
        if isinstance(created, str):
            # Parse ISO format string from Vault
            try:
                created = datetime.fromisoformat(created.replace("Z", "+00:00"))
            except (ValueError, TypeError):
                return AgeReport(
                    path=path,
                    current_version=metadata.version,
                    created_time=None,
                    max_age_days=max_age_days,
                    needs_rotation=False,
                    detail=f"Cannot parse created_time: {created}",
                )

        if created.tzinfo is None:
            created = created.replace(tzinfo=timezone.utc)

        delta = now - created
        age_days = delta.total_seconds() / 86400.0

    needs_rotation = age_days > max_age_days
    detail = ""
    if needs_rotation:
        overdue = age_days - max_age_days
        detail = f"Secret is {overdue:.1f} days overdue for rotation"
    elif age_days > max_age_days * 0.8:
        remaining = max_age_days - age_days
        detail = f"Secret will need rotation in {remaining:.1f} days"
    else:
        detail = "Secret age is within policy"

    return AgeReport(
        path=path,
        current_version=metadata.version,
        created_time=created if isinstance(created, datetime) else None,
        age_days=age_days,
        max_age_days=max_age_days,
        needs_rotation=needs_rotation,
        detail=detail,
    )


def check_secrets_batch(
    vault_client: "VaultClient",
    paths: list[str],
    policies: list[RotationPolicy] | None = None,
) -> list[AgeReport]:
    """Check multiple secrets against rotation policies.

    For each path, finds the matching policy (or uses default 90 days)
    and checks the secret age.

    Args:
        vault_client: Authenticated VaultClient instance.
        paths: List of secret paths to check.
        policies: Rotation policies to match against. Uses DEFAULT_POLICIES if None.

    Returns:
        List of AgeReport objects, one per path.
    """
    active_policies = policies or DEFAULT_POLICIES
    reports: list[AgeReport] = []

    for path in paths:
        max_age = 90.0  # default
        for policy in active_policies:
            if policy.matches_path(path):
                max_age = policy.max_age_days
                break
        try:
            report = check_secret_age(vault_client, path, max_age_days=max_age)
            reports.append(report)
        except RotationError as exc:
            reports.append(
                AgeReport(
                    path=path,
                    max_age_days=max_age,
                    needs_rotation=False,
                    detail=f"Error: {exc}",
                )
            )

    return reports


def rotate_sops_keys(
    repo_root: str | Path,
    new_key: str,
    key_type: str = "age",
    old_key: str = "",
    dry_run: bool = False,
) -> list[dict[str, str]]:
    """Rotate SOPS encryption keys across all encrypted files in a repository.

    This updates the .sops.yaml configuration and re-encrypts all matching
    files with the new key. Requires the old key to be available for
    decryption during the rotation.

    Args:
        repo_root: Repository root directory.
        new_key: New encryption key (age public key, KMS ARN, etc.).
        key_type: Key type ("age", "kms", "gcp_kms", "azure_keyvault").
        old_key: Old key to replace. If empty, the new key is added alongside existing keys.
        dry_run: If True, report what would change without modifying files.

    Returns:
        List of dicts with {"file": path, "status": "rotated"|"skipped"|"error", "detail": ...}.

    Raises:
        RotationError: If sops is not installed or repo_root is invalid.
    """
    root = Path(repo_root)
    if not root.is_dir():
        raise RotationError(f"Repository root not found: {root}")

    sops_yaml = root / ".sops.yaml"
    if not sops_yaml.exists():
        raise RotationError(f".sops.yaml not found in {root}")

    if shutil.which("sops") is None:
        raise RotationError("sops binary not found on PATH")

    results: list[dict[str, str]] = []

    # Step 1: Update .sops.yaml
    raw = sops_yaml.read_text()
    config = _load_yaml_preserving(raw)

    if not isinstance(config, dict) or "creation_rules" not in config:
        raise RotationError("Invalid .sops.yaml: missing creation_rules")

    updated_rules = False
    for rule in config["creation_rules"]:
        if not isinstance(rule, dict):
            continue

        current_value = rule.get(key_type, "")
        if old_key and current_value == old_key:
            if not dry_run:
                rule[key_type] = new_key
            updated_rules = True
        elif not old_key and key_type in rule:
            # Add new key alongside existing
            if not dry_run:
                rule[key_type] = new_key
            updated_rules = True

    if updated_rules and not dry_run:
        import yaml
        sops_yaml.write_text(yaml.safe_dump(config, default_flow_style=False))
        results.append({"file": str(sops_yaml), "status": "rotated", "detail": f"Updated {key_type} key"})
    elif updated_rules:
        results.append({"file": str(sops_yaml), "status": "would_rotate", "detail": f"Would update {key_type} key"})

    # Step 2: Find and re-encrypt all SOPS-encrypted files
    encrypted_files = _find_encrypted_files(root)

    for enc_file in encrypted_files:
        rel = enc_file.relative_to(root)
        if dry_run:
            results.append({"file": str(rel), "status": "would_rotate", "detail": "Would re-encrypt"})
            continue

        try:
            # sops updatekeys re-encrypts with the new .sops.yaml keys
            proc = subprocess.run(
                ["sops", "updatekeys", "--yes", str(enc_file)],
                capture_output=True,
                text=True,
                timeout=60,
            )
            if proc.returncode == 0:
                results.append({"file": str(rel), "status": "rotated", "detail": "Re-encrypted with new keys"})
            else:
                results.append({"file": str(rel), "status": "error", "detail": proc.stderr.strip()})
        except subprocess.TimeoutExpired:
            results.append({"file": str(rel), "status": "error", "detail": "sops updatekeys timed out"})
        except Exception as exc:
            results.append({"file": str(rel), "status": "error", "detail": str(exc)})

    return results


# ------------------------------------------------------------------
# Helpers
# ------------------------------------------------------------------

def _glob_to_regex(pattern: str) -> str:
    """Convert a simple glob pattern to a regex.

    Supports `*` as a single-segment wildcard.
    """
    parts = []
    for char in pattern:
        if char == "*":
            parts.append("[^/]*")
        elif char in r"\.+^${}()|[]":
            parts.append(f"\\{char}")
        else:
            parts.append(char)
    return "^" + "".join(parts) + "$"


def _find_encrypted_files(root: Path) -> list[Path]:
    """Find all SOPS-encrypted files in the repo."""
    encrypted: list[Path] = []
    for ext in ("*.enc.yaml", "*.enc.yml", "*.enc.json", "*.enc.env"):
        encrypted.extend(root.rglob(ext))
    return sorted(encrypted)


def _load_yaml_preserving(content: str) -> Any:
    """Load YAML content."""
    import yaml
    return yaml.safe_load(content)

File: lib/typescript/src/rotation.ts
/**
 * Secret rotation helpers.
 *
 * Provides rotation policy definitions, age checking for Vault secrets,
 * and batch checking utilities.
 */

import type { AgeReport } from "./models";
import { RotationError } from "./exceptions";
import type { VaultClient } from "./vault";

// ------------------------------------------------------------------
// Rotation Policy
// ------------------------------------------------------------------

export interface RotationPolicy {
  name: string;
  max_age_days: number;
  paths: string[];
  warn_age_days: number;
  auto_rotate: boolean;
  notify_channels: string[];
}

export function createRotationPolicy(
  name: string,
  opts: {
    max_age_days?: number;
    paths?: string[];
    warn_age_days?: number;
    auto_rotate?: boolean;
    notify_channels?: string[];
  } = {}
): RotationPolicy {
  const maxAge = opts.max_age_days ?? 90;
  return {
    name,
    max_age_days: maxAge,
    paths: opts.paths ?? [],
    warn_age_days: opts.warn_age_days ?? maxAge * 0.8,
    auto_rotate: opts.auto_rotate ?? false,
    notify_channels: opts.notify_channels ?? [],
  };
}

function globToRegex(pattern: string): RegExp {
  let regexStr = "^";
  for (const char of pattern) {
    if (char === "*") {
      regexStr += "[^/]*";
    } else if ("\\.+^${}()|[]".includes(char)) {
      regexStr += `\\${char}`;
    } else {
      regexStr += char;
    }
  }
  regexStr += "$";
  return new RegExp(regexStr);
}

export function policyMatchesPath(policy: RotationPolicy, secretPath: string): boolean {
  return policy.paths.some((pattern) => {
    const regex = globToRegex(pattern);
    return regex.test(secretPath);
  });
}

// ------------------------------------------------------------------
// Default Policies
// ------------------------------------------------------------------

export const DEFAULT_POLICIES: RotationPolicy[] = [
  createRotationPolicy("database-credentials", {
    max_age_days: 90,
    paths: ["kv/data/*/database/*", "kv/data/*/db/*"],
  }),
  createRotationPolicy("api-keys", {
    max_age_days: 180,
    paths: ["kv/data/*/api-keys/*", "kv/data/*/apikeys/*"],
  }),
  createRotationPolicy("service-accounts", {
    max_age_days: 365,
    paths: ["kv/data/*/service-accounts/*"],
  }),
  createRotationPolicy("tls-certificates", {
    max_age_days: 90,
    paths: ["kv/data/*/certs/*", "kv/data/*/tls/*"],
  }),
  createRotationPolicy("ssh-keys", {
    max_age_days: 90,
    paths: ["kv/data/*/ssh/*"],
  }),
];

// ------------------------------------------------------------------
// Age Checking
// ------------------------------------------------------------------

export async function checkSecretAge(
  vaultClient: VaultClient,
  secretPath: string,
  maxAgeDays: number = 90
): Promise<AgeReport> {
  try {
    const metadata = await vaultClient.kvMetadata(secretPath);

    let ageDays = 0;
    let created: Date | null = null;

    if (metadata.created_time) {
      try {
        created = new Date(metadata.created_time);
        if (isNaN(created.getTime())) {
          return {
            path: secretPath,
            current_version: metadata.version,
            created_time: null,
            age_days: 0,
            max_age_days: maxAgeDays,
            needs_rotation: false,
            detail: `Cannot parse created_time: ${metadata.created_time}`,
          };
        }
        ageDays = (Date.now() - created.getTime()) / (86400 * 1000);
      } catch {
        return {
          path: secretPath,
          current_version: metadata.version,
          created_time: null,
          age_days: 0,
          max_age_days: maxAgeDays,
          needs_rotation: false,
          detail: `Cannot parse created_time: ${metadata.created_time}`,
        };
      }
    }

    const needsRotation = ageDays > maxAgeDays;
    let detail: string;
    if (needsRotation) {
      const overdue = ageDays - maxAgeDays;
      detail = `Secret is ${overdue.toFixed(1)} days overdue for rotation`;
    } else if (ageDays > maxAgeDays * 0.8) {
      const remaining = maxAgeDays - ageDays;
      detail = `Secret will need rotation in ${remaining.toFixed(1)} days`;
    } else {
      detail = "Secret age is within policy";
    }

    return {
      path: secretPath,
      current_version: metadata.version,
      created_time: created?.toISOString() ?? null,
      age_days: ageDays,
      max_age_days: maxAgeDays,
      needs_rotation: needsRotation,
      detail,
    };
  } catch (err) {
    // VaultSecretNotFound
    if (err && typeof err === "object" && "name" in err && (err as { name: string }).name === "VaultSecretNotFound") {
      return {
        path: secretPath,
        current_version: 0,
        created_time: null,
        age_days: 0,
        max_age_days: maxAgeDays,
        needs_rotation: false,
        detail: `Secret not found at path: ${secretPath}`,
      };
    }
    // VaultConnectionError
    if (err && typeof err === "object" && "name" in err && (err as { name: string }).name === "VaultConnectionError") {
      throw new RotationError(`Cannot check age of ${secretPath}: ${err instanceof Error ? err.message : String(err)}`);
    }
    throw new RotationError(
      `Cannot check age of ${secretPath}: ${err instanceof Error ? err.message : String(err)}`
    );
  }
}

export async function checkSecretsBatch(
  vaultClient: VaultClient,
  paths: string[],
  policies?: RotationPolicy[]
): Promise<AgeReport[]> {
  const activePolicies = policies ?? DEFAULT_POLICIES;
  const reports: AgeReport[] = [];

  for (const secretPath of paths) {
    let maxAge = 90;
    for (const policy of activePolicies) {
      if (policyMatchesPath(policy, secretPath)) {
        maxAge = policy.max_age_days;
        break;
      }
    }

    try {
      const report = await checkSecretAge(vaultClient, secretPath, maxAge);
      reports.push(report);
    } catch (err) {
      reports.push({
        path: secretPath,
        current_version: 0,
        age_days: 0,
        max_age_days: maxAge,
        needs_rotation: false,
        detail: `Error: ${err instanceof Error ? err.message : String(err)}`,
      });
    }
  }

  return reports;
}
